### PR TITLE
Make TransportLocalClusterStateAction wait for cluster to unblock

### PR DESCRIPTION
**Pull request by nielsbauman**
This will make `TransportLocalClusterStateAction` wait for a new state
that is not blocked. This means we need a timeout (again). For
consistency's sake, we're reusing the REST param `master_timeout` for
this timeout as well.

The only class that was using `TransportLocalClusterStateAction` was
`TransportGetAliasesAction`, so its request needed to accept a timeout
again as well.

### CODE REVIEW

File: docs/reference/cat/alias.asciidoc [hunk 1]
```diff
@@ -51,6 +51,8 @@ include::{es-ref-dir}/rest-api/common-parms.asciidoc[tag=cat-v]
 
 include::{es-ref-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
 
+include::{es-ref-dir}/rest-api/common-parms.asciidoc[tag=master-timeout]
```
Comment (nielsbauman) [inline docs/reference/cat/alias.asciidoc:54]: Not sure if these docs and rest-api-spec changes are out of scope or not for this PR.

File: server/src/main/java/module-info.java [hunk 2]
```diff
@@ -146,6 +146,7 @@
     exports org.elasticsearch.action.support.master;
     exports org.elasticsearch.action.support.master.info;
     exports org.elasticsearch.action.support.nodes;
+    exports org.elasticsearch.action.support.local;
```
Comment (nielsbauman) [inline server/src/main/java/module-info.java:149]: I moved `TransportLocalClusterStateAction` to its own `local` package - together with the new `LocalClusterStateRequest`.

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java [hunk 3]
```diff
@@ -31,18 +30,24 @@ public class GetAliasesRequest extends ActionRequest implements AliasesRequest {
     private String[] indices = Strings.EMPTY_ARRAY;
     private IndicesOptions indicesOptions = DEFAULT_INDICES_OPTIONS;
 
+    @Deprecated
     public GetAliasesRequest(String... aliases) {
-        this.aliases = aliases;
-        this.originalAliases = aliases;
+        this(MasterNodeRequest.TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT, aliases);
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:35]: It felt out of scope for this PR to refactor all these constructor callers to provide an explicit master timeout - that will have to be done in a follow-up PR.

File: server/src/main/java/org/elasticsearch/action/support/local/LocalClusterStateRequest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A base request for actions that are executed locally on the node that receives the request.
+ */
+public abstract class LocalClusterStateRequest extends ActionRequest {
+
+    private final TimeValue clusterUpdateTimeout;
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/action/support/local/LocalClusterStateRequest.java:25]: I was on the fence about naming this `clusterUpdateTimeout` or `masterTimeout`. Opinions are welcome :)

**Comment by nielsbauman** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/support/local/LocalClusterStateRequest.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/local/LocalClusterStateRequest.java:25]: Let's call it `masterTimeout` to align with the REST query parameter. If you're uncomfortable with the name, maybe add Javadoc describing what it means and how it differs from `MasterNodeRequest#masterTimeout`.

File: server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    protected final void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final var state = clusterService.state();
+        final var clusterBlockException = checkBlock(request, state);
+        if (clusterBlockException != null) {
+            if (clusterBlockException.retryable() == false) {
+                listener.onFailure(clusterBlockException);
+            } else {
+                waitForClusterUnblock(task, request, listener, state, clusterBlockException);
+            }
+            return;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java:75]: Let's just use an `else` here rather than an early return.

File: server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.debug(
+                    () -> format("timed out while waiting for cluster to unblock in [%s] (timeout [%s])", actionName, timeout),
+                    exception
+                );
+                listener.onFailure(new ElasticsearchTimeoutException("timed out while waiting for cluster to unblock", exception));
+            }
+        }, clusterState -> isTaskCancelled(task) || checkBlock(request, clusterState) == null);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java:123]: This means that if the task is cancelled while waiting for an unblocking cluster state then it'll wait for the next cluster state update before completing, which could be arbitrarily far in the future. Could we use `org.elasticsearch.tasks.CancellableTask#addListener` to react more promptly to cancellation instead?

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:35]: I would rather we did those changes first to set things up for this PR to go through without having to add back this trappy parameter.

File: docs/reference/cat/alias.asciidoc (hunk 1 above)
Comment (DaveCTurner) [reply docs/reference/cat/alias.asciidoc:54]: Seems ok to me, tho note that we have never before supported `?master_timeout` on these APIs, this isn't reinstating a parameter that previously was removed.

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java (hunk 3 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:35]: How would we do those changes first if there is currently - on `main` - no timeout to set in these constructors? Are you suggesting adding a constructor on `main` that takes a timeout but doesn't actually do anything with it and swapping that constructor with the ones on this branch afterward?

File: docs/reference/cat/alias.asciidoc (hunk 1 above)
Comment (nielsbauman) [reply docs/reference/cat/alias.asciidoc:54]: Yeah I realized that, but I figured it made sense to add them if they're parameters we're accepting.

File: server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java (hunk 6 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java:123]: Good idea :+1:. It looks like we're not doing that in `TransportMasterNodeAction` either. Am I missing something or should we add it there too?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java:123]: Eh we probably should. But that's a good point, let's leave this with the same behaviour as in `TransportMasterNodeAction` and we can come back to this change later.

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java [hunk 7]
```diff
@@ -31,18 +30,24 @@ public class GetAliasesRequest extends ActionRequest implements AliasesRequest {
...
 
+    @Deprecated
     public GetAliasesRequest() {
         this(Strings.EMPTY_ARRAY);
     }
 
-    @Override
-    public void writeTo(StreamOutput out) throws IOException {
-        TransportAction.localOnly();
+    public GetAliasesRequest(TimeValue clusterUpdateTimeout, String... aliases) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:43]: Naming nit: `masterTimeout` rather than `clusterUpdateTimeout`

File: server/src/test/java/org/elasticsearch/action/support/local/TransportLocalClusterStateActionTests.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        assertTrue(exception.getCause() instanceof ClusterBlockException);
+    }
+
+    public void testTaskCancelledAfterBlock() {
+        var request = new Request();
+        ClusterBlock block = new ClusterBlock(1, "", true, true, false, randomFrom(RestStatus.values()), ClusterBlockLevel.ALL);
+        var state = ClusterState.builder(clusterService.state()).blocks(ClusterBlocks.builder().addGlobalBlock(block)).build();
+        setState(clusterService, state);
+
+        Task task = new CancellableTask(randomLong(), "test", Action.ACTION_NAME, "", TaskId.EMPTY_TASK_ID, Map.of());
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/action/support/local/TransportLocalClusterStateActionTests.java:140]: nit (avoids the need to cast it below)

```suggestion
        CancellableTask task = new CancellableTask(randomLong(), "test", Action.ACTION_NAME, "", TaskId.EMPTY_TASK_ID, Map.of());
```

**Review by DaveCTurner — COMMENTED** [review]
Looks good, just a few superficial things

File: docs/changelog/117230.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+pr: 117230
+summary: Make `TransportLocalClusterStateAction` wait for cluster to unblock
```
[the remark this answers is not in the document]
Comment (nielsbauman) [reply docs/changelog/117230.yaml]: I changed the summary to `Make various alias retrieval APIs wait for cluster to unblock`. Let me know if you have any suggestions and/or if you think we should elaborate with a description.

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java (hunk 6 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java:122]: I opened https://github.com/elastic/elasticsearch/issues/117971 for this.
